src/backend/distributed/utils/aggregate_utils.c

```diff
@@ -0,0 +1,590 @@
+	}
+	PG_RETURN_DATUM(result);
+}
+
+
+/*
+ * (box, agg, text) -> box
+ * box.agg = agg
+ * box.value = agg.combine(box.value, agg.stype.input(text))
+ * return box
+ */
+Datum
+coord_combine_agg_sfunc(PG_FUNCTION_ARGS)
+{
+	LOCAL_FCINFO(inner_fcinfo, 3);
+	FmgrInfo info;
+	HeapTuple aggtuple;
+	HeapTuple transtypetuple;
+	Form_pg_aggregate aggform;
+	Form_pg_type transtypeform;
+	Oid combine;
+	Oid deserial;
+	Oid ioparam;
+	Datum value;
+	bool value_null;
+	StypeBox *box;
+
+	if (PG_ARGISNULL(0))
+	{
+		box = pallocInAggContext(fcinfo, sizeof(StypeBox));
+		box->agg = PG_GETARG_OID(1);
+	}
+	else
+	{
+		box = (StypeBox *) PG_GETARG_POINTER(0);
+		Assert(box->agg == PG_GETARG_OID(1));
+	}
+
+	aggtuple = get_aggform(box->agg, &aggform);
+
+	if (aggform->aggcombinefn == InvalidOid)
+	{
+		elog(ERROR, "coord_combine_agg_sfunc expects an aggregate with COMBINEFUNC.");
+	}
+
+	if (aggform->aggtranstype == INTERNALOID)
+	{
+		elog(ERROR,
+			 "coord_combine_agg_sfunc does not support aggregates with INTERNAL transition state,");
+	}
+
+	combine = aggform->aggcombinefn;
+
+	if (PG_ARGISNULL(0))
+	{
+		InitializeStypeBox(fcinfo, box, aggtuple, aggform->aggtranstype);
+	}
+
+	ReleaseSysCache(aggtuple);
+
+	if (PG_ARGISNULL(0))
+	{
+		get_typlenbyval(box->transtype,
+						&box->transtypeLen,
+						&box->transtypeByVal);
+	}
+
+	value_null = PG_ARGISNULL(2);
+	transtypetuple = get_typeform(box->transtype, &transtypeform);
+	ioparam = getTypeIOParam(transtypetuple);
+	deserial = transtypeform->typinput;
+	ReleaseSysCache(transtypetuple);
+
+	fmgr_info(deserial, &info);
+	if (value_null && info.fn_strict)
+	{
+		value = (Datum) 0;
+	}
+	else
+	{
+		InitFunctionCallInfoData(*inner_fcinfo, &info, 3, fcinfo->fncollation,
+								 fcinfo->context, fcinfo->resultinfo);
+		fcSetArgExt(inner_fcinfo, 0, PG_GETARG_DATUM(2), value_null);
+		fcSetArg(inner_fcinfo, 1, ObjectIdGetDatum(ioparam));
+		fcSetArg(inner_fcinfo, 2, Int32GetDatum(-1)); /* typmod */
+
+		value = FunctionCallInvoke(inner_fcinfo);
+		value_null = inner_fcinfo->isnull;
+	}
+
+	fmgr_info(combine, &info);
+
+	if (info.fn_strict)
+	{
+		if (value_null)
+		{
+			PG_RETURN_POINTER(box);
+		}
+
+		if (!box->value_init)
+		{
+			agg_sfunc_handle_strict_uninit(box, fcinfo, value);
+			PG_RETURN_POINTER(box);
+		}
+
+		if (box->value_null)
+		{
+			PG_RETURN_POINTER(box);
+		}
+	}
+
+	InitFunctionCallInfoData(*inner_fcinfo, &info, 2, fcinfo->fncollation,
+							 fcinfo->context, fcinfo->resultinfo);
+	fcSetArgExt(inner_fcinfo, 0, box->value, box->value_null);
+	fcSetArgExt(inner_fcinfo, 1, value, value_null);
+
+	agg_sfunc_handle_transition(box, fcinfo, inner_fcinfo);
+
+	PG_RETURN_POINTER(box);
+}
+
+
+/*
+ * (box, ...) -> fval
+ * return box.agg.ffunc(box.value)
+ */
+Datum
+coord_combine_agg_ffunc(PG_FUNCTION_ARGS)
+{
+	Datum ret;
+	StypeBox *box = (StypeBox *) (PG_ARGISNULL(0) ? NULL : PG_GETARG_POINTER(0));
+	LOCAL_FCINFO(inner_fcinfo, FUNC_MAX_ARGS);
+	FmgrInfo info;
+	int inner_nargs;
```

Comment:
Should follow style guide in terms of naming, etc.

https://docs.google.com/document/d/104XKGvh8Z_AamjtUEmbzNKgx0F6qsKyqIOJD_QRDJzI/edit?usp=sharing

Reply:
> Should follow style guide in terms of naming, etc.
> 
> https://docs.google.com/document/d/104XKGvh8Z_AamjtUEmbzNKgx0F6qsKyqIOJD_QRDJzI/edit?usp=sharing

@serprex what was the resolution on this?

Reply:
Bump